File: DFS,BFS/11724.py
from collections import deque

N, M = map(int, input().split())
graph = [[] for _ in range(N + 1)]
visited = [False] * (N + 1)
cnt = 0

def bfs(graph,visited, start):
  queue = deque()
  queue.append(start)
  visited[start] = True
  while queue:
    v = queue.popleft()
    for i in graph[v]:
      if not visited[i]:
        visited[i] = True
        queue.append(i)

for i in range(M):
  start, end = map(int, input().split())
  graph[start].append(end)
  graph[end].append(start)

for i in range(1, len(visited)):
  if visited[i] == False:
    bfs(graph, visited, i)
    cnt += 1

print(cnt)

File: DFS,BFS/11724.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <bits/stdc++.h>
#define MAX 1001

using namespace std;
vector<int> graph[MAX];
bool visited[MAX];

void bfs(int start) {
	queue<int> q;
	q.push(start);
	visited[start] = true;

	while (!q.empty()) {
		int x = q.front();
		q.pop();
		for (int i = 0; i < graph[x].size(); i++) {
			int y = graph[x][i];
			if (!visited[y]) {
				visited[y] = true;
				q.push(y);
			}
		}
	}
}

int main(void) {
	int n, m;
	int count = 0;
	scanf("%d %d", &n, &m);

	for (int i = 0; i < m; i++) {
		int a, b;
		scanf("%d %d", &a, &b);
		graph[a].push_back(b);
		graph[b].push_back(a);
	}

	for (int i = 1; i < n+1; i++) {
		sort(graph[i].begin(), graph[i].end());
	}

	for (int i = 1; i < n + 1; i++) {
		if (!visited[i]) {
			bfs(i);
			count += 1;
		}
	}

	printf("%d", count);
	return 0;
}
